core/coreapi/swarm.go

```diff
@@ -0,0 +1,180 @@
+package coreapi
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"time"
+
+	coreiface "github.com/ipfs/go-ipfs/core/coreapi/interface"
+
+	swarm "gx/ipfs/QmPQoCVRHaGD25VffyB7DFV5qP65hFSQJdSDy75P1vYBKe/go-libp2p-swarm"
+	iaddr "gx/ipfs/QmSzdvo9aPzLj4HXWTcgGAp8N84tZc8LbLmFZFwUb1dpWk/go-ipfs-addr"
+	ma "gx/ipfs/QmYmsdtJ3HsodkePE3eU3TsCaP2YvPZJ4LoXnNkDE5Tpt7/go-multiaddr"
+	protocol "gx/ipfs/QmZNkThpqfVXs9GNbexPrfBbXSLNYeKrE7jwFM2oqHbyqN/go-libp2p-protocol"
+	peer "gx/ipfs/QmbNepETomvmXfz1X5pHNFD2QuPqnqi47dTd94QJWSorQ3/go-libp2p-peer"
+	pstore "gx/ipfs/QmfAQMFpgDU2U4BXG64qVr8HSiictfWvkSBz7Y2oDj65st/go-libp2p-peerstore"
+	inet "gx/ipfs/QmfDPh144WGBqRxZb1TGDHerbMnZATrHZggAPw7putNnBq/go-libp2p-net"
+	net "gx/ipfs/QmfDPh144WGBqRxZb1TGDHerbMnZATrHZggAPw7putNnBq/go-libp2p-net"
+)
+
+type SwarmAPI struct {
+	*CoreAPI
+}
+
+type connInfo struct {
+	api  *CoreAPI
+	conn net.Conn
+	dir  net.Direction
+
+	addr  ma.Multiaddr
+	peer  peer.ID
+	muxer string
+}
+
+func (api *SwarmAPI) Connect(ctx context.Context, pi pstore.PeerInfo) error {
+	if api.node.PeerHost == nil {
+		return coreiface.ErrOffline
+	}
+
+	swrm, ok := api.node.PeerHost.Network().(*swarm.Swarm)
+	if !ok {
+		return fmt.Errorf("peerhost network was not swarm")
```

Comment:
If possible, can we just make this optional? That is, clear the backoff iff we *have* a swarm but don't error out.

Reply:
(I realize this probably isn't what we do now but, well, no time better than now)